File: projecteuler.fsharp/ProjectEulerTasks/Problem074.fs
namespace ProjectEulerTasks

open NUnit.Framework
open ProjectEulerTasks.Utils
open CommonLib

// The number 145 is well known for the property that the sum of the factorial of its digits is equal to 145:
// 1! + 4! + 5! = 1 + 24 + 120 = 145
// Perhaps less well known is 169, in that it produces the longest chain of numbers that link back to 169;
// it turns out that there are only three such loops that exist:
// 169 -> 363601 -> 1454 -> 169
// 871 -> 45361 -> 871
// 872 -> 45362 -> 872
// It is not difficult to prove that EVERY starting number will eventually get stuck in a loop. For example,
// 69 -> 363600 -> 1454 -> 169 -> 363601 (-> 1454)
// 78 -> 45360 -> 871 -> 45361 (-> 871)
// 540 -> 145 (-> 145)
// Starting with 69 produces a chain of five non-repeating terms, but the longest non-repeating chain with a starting number below one million is sixty terms.
// How many chains, with a starting number below one million, contain exactly sixty non-repeating terms?

module Problem074Impl =

    type ChainResult =
    | LoopDetected of current: int * loopSize: int * chain: int list
    | SavedValueDetected of current: int * loopSize: int * chain: int list

open Problem074Impl

[<TestFixture>]
type Problem074() =

    let digitFactorials = seq { 0 .. 9 } |> Seq.map (fun digit -> digit |> Numbers.CalcFactorial |> int) |> Seq.toArray

    let generateNextStep (current: int) =
        current |> NumbersDigits.GetDigits |> Seq.map (fun digit -> digitFactorials.[digit]) |> Seq.sum

    let getChainLoop (current: int) (chain: int list) =
        match chain |> List.tryFindIndex (fun number -> number = current) with
        | Some index -> index + 1 |> Some
        | None -> None

    let saveChain (storage: SafeStorage<int>) (result: ChainResult) =
        match result with
        | ChainResult.LoopDetected (current, loopSize, chain) ->
            chain |> Seq.take loopSize |> Seq.iter (fun number -> storage.SetValue(number, loopSize))
            chain |> Seq.skip loopSize |> Seq.iteri (fun index number -> storage.SetValue(number, loopSize + index + 1))
        | ChainResult.SavedValueDetected (current, loopSize, chain) ->
            chain |> Seq.iteri (fun index number -> storage.SetValue(number, loopSize + index + 1))

    let rec processChain (storage: SafeStorage<int>) (chain: int list) (current: int) =
        match current |> storage.GetValue with
        | value when value = storage.DefaultValue ->
            match getChainLoop current chain with
            | Some loopSize -> ChainResult.LoopDetected (current, loopSize, chain)
            | None -> current |> generateNextStep |> processChain storage (current :: chain)
        | value -> ChainResult.SavedValueDetected (current, value, chain)

    let addNumber (number: int) (storage: SafeStorage<int>) =
        number |> processChain storage [] |> saveChain storage

    let solveImpl (maxNumber: int) (expectedChainSize: int) =
        let storage = SafeStorage(1, maxNumber, 0)
        seq { storage.MinNumber .. storage.MaxNumber } |> Seq.iter (fun number -> addNumber number storage)
        storage.Storage |> Seq.filter (fun value -> value = expectedChainSize) |> Seq.length

    [<TestCase(999999, 60, 402, TimeThresholds.HardTimeLimit)>]
    member public this.Solve(maxNumber: int, expectedChainSize: int, expectedAnswer: int, timeLimit: int) =
        SolutionUtils.CheckSolution(timeLimit, expectedAnswer, solveImpl, maxNumber, expectedChainSize)

File: projecteuler.erlang/project_euler/src/problem_074.erl
%% @author std-string

%% The number 145 is well known for the property that the sum of the factorial of its digits is equal to 145:
%% 1! + 4! + 5! = 1 + 24 + 120 = 145
%% Perhaps less well known is 169, in that it produces the longest chain of numbers that link back to 169;
%% it turns out that there are only three such loops that exist:
%% 169 -> 363601 -> 1454 -> 169
%% 871 -> 45361 -> 871
%% 872 -> 45362 -> 872
%% It is not difficult to prove that EVERY starting number will eventually get stuck in a loop. For example,
%% 69 -> 363600 -> 1454 -> 169 -> 363601 (-> 1454)
%% 78 -> 45360 -> 871 -> 45361 (-> 871)
%% 540 -> 145 (-> 145)
%% Starting with 69 produces a chain of five non-repeating terms, but the longest non-repeating chain with a starting number below one million is sixty terms.
%% How many chains, with a starting number below one million, contain exactly sixty non-repeating terms?

-module(problem_074).
-export([get_check_data/0, prepare_data/2, solve/1]).

-behaviour(numerical_task_behaviour).

-type factorial_storage() :: array:array(numbers:digit()).
-type chain_storage() :: array:array('undef' | {'index', Index :: non_neg_integer()} | pos_integer()).
-type number_search_result() :: {'true', NextNumber :: pos_integer()} | 'false'.
-type result() :: {'index', CycleStart :: pos_integer(), CycleSize :: pos_integer()} | pos_integer().

%% ====================================================================
%% API functions
%% ====================================================================

-spec get_check_data() -> [{Input :: term(), Output :: term()}].
get_check_data() -> [{{1000000 - 1, 60}, 402}].

-spec prepare_data(ModuleSourceDir :: string(), Input :: term()) -> term().
prepare_data(_ModuleSourceDir, Input) -> Input.

-spec solve(PreparedInput :: term()) -> term().
solve({MaxNumber, ExpectedChainSize}) ->
    FactorialStorage = array:from_list(lists:map(fun(Digit) -> numbers:factorial(Digit) end, lists:seq(0, 9))),
    StorageSize = calc_storage_size(MaxNumber),
    InitStorage = array:new([{size, StorageSize}, {fixed, true}, {default, undef}]),
    Storage = process_number(1, MaxNumber, InitStorage, FactorialStorage),
    process_storage(0, MaxNumber, Storage, ExpectedChainSize, 0).

%% ====================================================================
%% Internal functions
%% ====================================================================

-spec calc_storage_size(MaxNumber :: pos_integer()) -> pos_integer().
calc_storage_size(MaxNumber) ->
    %% rough solution
    length(numbers:get_digits(MaxNumber)) * numbers:factorial(9).

-spec process_storage(Index :: non_neg_integer(),
                      MaxNumber :: pos_integer(),
                      Storage :: chain_storage(),
                      ExpectedChainSize :: pos_integer(),
                      Count :: pos_integer()) -> pos_integer().
process_storage(MaxNumber, MaxNumber, _Storage, _ExpectedChainSize, Count) -> Count;
process_storage(Index, MaxNumber, Storage, ExpectedChainSize, Count) ->
    case array:get(Index, Storage) of
        ExpectedChainSize -> process_storage(Index + 1, MaxNumber, Storage, ExpectedChainSize, Count + 1);
        _Other -> process_storage(Index + 1, MaxNumber, Storage, ExpectedChainSize, Count)
    end.

-spec process_number(Number :: pos_integer(),
                     MaxNumber :: pos_integer(),
                     Storage :: chain_storage(),
                     FactorialStorage :: factorial_storage()) -> chain_storage().
process_number(Number, MaxNumber, Storage, FactorialStorage) ->
    NewStorage = process_chain(Number, [], 0, Storage, FactorialStorage),
    case find_next_number(Number + 1, MaxNumber, Storage) of
        {true, NextNumber} -> process_number(NextNumber, MaxNumber, NewStorage, FactorialStorage);
        false -> NewStorage
    end.

-spec find_next_number(Number :: pos_integer(), MaxNumber :: pos_integer(), Storage :: chain_storage()) -> number_search_result().
find_next_number(Number, MaxNumber, _Storage) when Number > MaxNumber -> false;
find_next_number(Number, MaxNumber, Storage) ->
    case array:get(Number - 1, Storage) of
        undef -> {true, Number};
        _Other -> find_next_number(Number + 1, MaxNumber, Storage)
    end.

-spec process_chain(Number :: pos_integer(),
                    Chain :: [pos_integer()],
                    ChainIndex :: non_neg_integer(),
                    Storage :: chain_storage(),
                    FactorialStorage :: factorial_storage()) -> chain_storage().
process_chain(Number, Chain, ChainIndex, Storage, FactorialStorage) ->
    case array:get(Number - 1, Storage) of
        undef ->
            NewStorage = array:set(Number - 1, {index, ChainIndex}, Storage),
            process_chain(process_step(Number, FactorialStorage), [Number] ++ Chain, ChainIndex + 1, NewStorage, FactorialStorage);
        {index, Index} -> collect_result(Chain, {index, Number, ChainIndex - Index}, Storage);
        Count -> collect_result(Chain, Count + 1, Storage)
    end.

-spec collect_result(Chain :: [pos_integer()], Result :: result(), Storage :: chain_storage()) -> chain_storage().
collect_result([], _Result, Storage) -> Storage;
collect_result([CycleStart | ChainRest], {index, CycleStart, CycleSize}, Storage) ->
    collect_result(ChainRest, CycleSize + 1, array:set(CycleStart - 1, CycleSize, Storage));
collect_result([Number | ChainRest], {index, CycleStart, CycleSize}, Storage) ->
    collect_result(ChainRest, {index, CycleStart, CycleSize}, array:set(Number - 1, CycleSize, Storage));
collect_result([Number | ChainRest], Count, Storage) ->
    collect_result(ChainRest, Count + 1, array:set(Number - 1, Count, Storage)).

-spec process_step(Number :: pos_integer(), FactorialStorage :: factorial_storage()) -> pos_integer().
process_step(Number, FactorialStorage) ->
    lists:foldl(fun(Digit, Sum) -> array:get(Digit, FactorialStorage) + Sum end, 0, numbers:get_digits(Number)).